[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations

mrzzd opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46493 [fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations**

This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

Commit:
[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations

This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

[ghstack-poisoned]

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -119,34 +113,31 @@ def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> boo
         """
         return m.__module__.startswith('torch.nn') and not isinstance(m, torch.nn.Sequential)
 
+    def get_module_qualified_name(self, mod):
```
Can you describe the case where this logic doesn't work and needs to be overridden? I want to understand more about why we are not finding the module by searching the module hierarchy for it.

zdevito replied:
nit: `path_of_module` might be a better name because torch.nn.Linear.__qualname__ exists and can be mistaken for the qualified name of a module..

mrzzd replied:
> Can you describe the case where this logic doesn't work and needs to be overridden? I want to understand more about why we are not finding the module by searching the module hierarchy for it.

An example is internal dper3 code, where modules are wrapped by PyTorchModuleWrapper. Please see how this function is reimplemented in D24239120.

zdevito commented on torch/fx/symbolic_trace.py:
```diff
@@ -119,34 +113,31 @@ def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> boo
+                return n
+        raise NameError('module is not installed as a submodule')
+
     def call_module(self, m: torch.nn.Module, module_qualified_name: str, forward: Callable[..., Any], args, kwargs):
```
once `get_module_qualified_name` is public then `call_module` doesn't need to take `module_qualified_name` as an argument, which would simplify things.

zdevito reviewed (approved):
Cool!  Some minor comments are inline.

Commit:
Update on "[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations"


This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

[ghstack-poisoned]

Commit:
Update on "[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations"


This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

[ghstack-poisoned]

Commit:
Update on "[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations"


This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

[ghstack-poisoned]

Commit:
Update on "[fx] Refactor Tracer so that find_module and root args creation could be overridden by implementations"


This will allow us to override tow following methods of Tracer:
-- get_module_qualified_name: to find qualified name of a module. In default implementation, it looks for module in registered modules and from there it gets to the name, but in some scenarios, the module being called could not be the exact same module that was registered.
-- create_args_for_root: This allows to create and pass custom structured input (like dictionary with specific keys) to the main module, rather than pure proxy objects. This will also allows us to let proxy objects only represent tensors when they are passed to modules.

Differential Revision: [D24269034](https://our.internmc.facebook.com/intern/diff/D24269034/)

[ghstack-poisoned]

malfet commented:
Reverting, as it caused number of CI failures, for example:
https://app.circleci.com/pipelines/github/pytorch/pytorch/228804/workflows/d978464a-8e6b-49eb-a422-d8de49b18fe4/jobs/8324375
